manage

```diff
@@ -0,0 +1,593 @@
+  OK        : test if virtualenv is OK
+pybuild :
+  Build python packages at ./${PYDIST}
+pypi.upload:
+  Upload python packages to PyPi (to test use pypi.upload.test)
+pyclean :
+  delete virtualenv and intermediate py files
+buildenv :
+  run pyenv.assert first and afterwards ./utils/brand.env is build
+test.* :
+  pylint    : lint PYLINT_FILES, searx/engines, searx & tests
+  pep8      : pycodestyle (pep8) for all files except PYLINT_FILES
+  unit      : run unit tests
+  coverage  : run unit tests with coverage
+  robot     : run robot test
+  clean     : clean intermediate test stuff
+node.* :
+  env       : download & install npm dependencies locally
+  clean     : drop npm installations
+themes.* :
+  oscar     : build oscar theme
+  simple    : build simple theme
+  bootstrap : less compile bootstrap.min.css CSS
+engines.languages :
+  update searx/data/useragents.json and searx/languages.py
+babel.compile :
+  pybabel compile ./searx/translations
+useragents.update :
+  update searx/data/useragents.json with the most recent versions of Firefox.
+docs.* :
+  html      : build HTML documentation
+  autobuild : autobuild HTML documentation while editing
+  prebuild  : build reST include files (./${DOCS_BUILD}/includes)
+  clean     : clean documentation build
+docker.build [push] :
+  build (and push) docker image
+gecko.driver :
+  download & install geckodriver if not already installed (required for
+  robot_tests)
+EOF
+    [[ -n ${1} ]] && err_msg "$cmd"
+    return 0
+}
+
+
+# shellcheck disable=SC2119
+main() {
+
+    local err=0
+    local cmd="$1"; shift
+
+    case "$cmd" in
+        --getenv) var="$1"; echo "${!var}"; exit 0;;
+        ci.*)
+            (   set -e
+                "$cmd" "$@"
+            )
+            dump_return $?
+            ;;
+        pyenv.assert|pyenv.cmd|build_msg)
+            "$cmd" "$@"
+            ;;
+        pyenv.OK|pyclean|pybuild|pypi.upload*)
+            "$cmd"
+            ;;
+        buildenv)
+            "$cmd"
+            ;;
+        test.pylint|test.pep8|test.unit|test.coverage|test.robot|test.clean)
+            "$cmd"
+            ;;
+        node.env|node.clean)
+            "$cmd"
+            ;;
+        themes.oscar|themes.simple|themes.bootstrap)
+            "$cmd"
+            ;;
+        engines.languages|babel.compile|useragents.update)
+            "$cmd"
+            ;;
+        docs.html|docs.autobuild)
+            docs.prebuild
+            "$cmd"
+            ;;
+        docs.prebuild|docs.clean)
+            "$cmd"
+            ;;
+        docker.build|gecko.driver)
+            "$cmd"
+            ;;
+        --help|help)
+            usage
+            ;;
+        *)
+            usage "unknown or missing command $1"
+            exit 42
+            ;;
+    esac
+}
+
+dump_return() {
+    local err=$1
+    [ "$err" -ne "0" ] && err_msg "${FUNCNAME[1]} exit with error ($err)"
+    return "$err"
+}
+
+# shellcheck disable=SC2120
+pyenv.assert() {
```

Comment:
> There is another implementation line 659 of utils/lib.sh

That's right. For me `./lib` is _abstract_ implementing basic stuff. `./lib.sh` is sourced by a _script_ and the functions from can be overwritten in the _script_ to fit the needs of the _application_ the script serves.

Here the _assert_ function calls additional `buildenv` compared to the _assert_ from lib.  This _assert_ also lets die the script if there are any problems when building (e.g.) the buildenv.

I know that you like to have a _straight forward_ implementation but a t this point I would say a minimal _abstraction_ is helpful.  The `./lib.sh` I also use in other projects (I maintain this lib since years) .. take it like any other lib we have in our project (e.g. github-pages-deploy-action in the build or jquery in the client).

Hope it is OK for you.

Reply:
The `pyenv.assert` function is no longer needed in the new implementation.